language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/references/ReferencesUtil.java

```diff
@@ -140,4 +107,45 @@ public static Range getRange(Location referencePos) {
                 referencePos.lineRange().endLine().line(), referencePos.lineRange().endLine().offset());
         return new Range(start, end);
     }
+    
+    /**
+     * Returns the symbol at cursor handling a special case where cursor position is at the RHS end of the symbol.
+     *
+     * @param context Operation context triggered with a cursor position
+     * @return Symbol at cursor
+     */
+    public static Optional<Symbol> getSymbolAtCursor(PositionedOperationContext context) {
+        Optional<Document> srcFile = context.currentDocument();
+        Optional<SemanticModel> semanticModel = context.currentSemanticModel();
+
+        if (semanticModel.isEmpty() || srcFile.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Position position = context.getCursorPosition();
+        Optional<Symbol> symbolAtCursor = semanticModel.get().symbol(srcFile.get(),
+                LinePosition.from(position.getLine(), position.getCharacter()));
+
+        if (symbolAtCursor.isEmpty()) {
+            if (position.getCharacter() == 0) {
+                return Optional.empty();
+            }
+
+            // If we did not find the symbol, there are 2 possibilities.
+            //  1. Cursor is at the end (RHS) of the symbol
+            //  2. Semantic API has a limitation
+            // Out of those, 2nd one is ignored assuming semantic API behaves correctly. 1st one is caused due to the
+            // right end column being excluded when searching. To overcome that, here we search for the symbol at 
+            // (col - 1). Ideally this shouldn't be an issue, because if we had cursor at the start col or middle, the
+            // 1st search (above) would have found that.
+            position = new Position(position.getLine(), position.getCharacter() - 1);
+            symbolAtCursor = semanticModel.get().symbol(srcFile.get(),
+                    LinePosition.from(position.getLine(), position.getCharacter()));
```

Comment:
Shall we get the position in 141 straightaway at line 143? then we don't have to call `getLine()` and `getCharacter()` fpr the second time.

Reply:
Done